### ISSUE TITLE
does tiny-dnn need to explicitly point out the size of the input and output of a convolutional layer?

### ISSUE THREAD

**Issue by Cv9527** [tiny-dnn/tiny-dnn#508]

**Comment by beru** [comment]
Almost so with standard approach since constructor methods of many layer classes take those size arguments. I think they should be ommitable (they can be set after NN is constructed) but it takes extra efforts to implement such feature.

**Comment by Cv9527** [comment]
does tiny-dnn can be used for fully cnn with multi-scale input images?

**Comment by edgarriba** [comment]
yes we support deconvolutional (or conv2d transposed) layers. For multi scale not sure what type of architecture do you mean here

**Comment by beru** [comment]
@edgarriba I'm not sure why you mentioned deconvolutional (or conv2d transposed) layers in here. Can you please explain it to me?

**Comment by beru** [comment]
If possible, input dimensions should be able to be set automatically by network class after instantiating layer classes. Below is the modified network definition of `examples/mnist/train.cpp`.

```cpp
nn << convolutional_layer<tan_h>(5, 6,             // C1, 1@32x32-in, 6@28x28-out
        padding::valid, true, 1, 1, backend_type)
   << average_pooling_layer<tan_h>(2)              // S2, 6@28x28-in, 6@14x14-out
   << convolutional_layer<tan_h>(5, 16,            // C3, 6@14x14-in, 16@10x10-out
        connection_table(tbl, 6, 16),
        padding::valid, true, 1, 1, backend_type)
   << average_pooling_layer<tan_h>(2)              // S4, 16@10x10-in, 16@5x5-out
   << convolutional_layer<tan_h>(5, 120,           // C5, 16@5x5-in, 120@1x1-out
        padding::valid, true, 1, 1, backend_type)
   << fully_connected_layer<tan_h>(10,             // F6, 120-in, 10-out
        true, backend_type)
;
```

And isn't it strange to set `backend_type` per layers?

**Comment by edgarriba** [comment]
@beru [FCNN](https://people.eecs.berkeley.edu/~jonlong/long_shelhamer_fcn.pdf) replace the final fully_connected layer (one prediction per class) by a deconvolution or more accurately a transposed conv2d in order to have a prediction for each output pixel. In the literature sometimes is referred as an up-sampling operation.

Check how it can be implemented using TF
http://stackoverflow.com/questions/37065512/fully-convolution-net-fcn-on-tensorflow

In addtion, check different SOTA convolution arithmetics
https://github.com/vdumoulin/conv_arithmetic

**Comment by bhack** [comment]
There are also some others interesting solutions like https://github.com/titu1994/Fully-Connected-DenseNets-Semantic-Segmentation

**Comment by beru** [comment]
@edgarriba , @bhack Thank you for the explanation and information.

Aside from fully cnn with multi-scale input images, don't you think it is a bit cumbersome to specify dimension parameters explicitly?

**Comment by bhack** [comment]
Probably
